fabric-component-access-api-v1/src/main/java/net/fabricmc/fabric/impl/component/access/ComponentTypeImpl.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.component.access;
+
+import java.util.IdentityHashMap;
+import java.util.function.BiPredicate;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import com.mojang.datafixers.util.Pair;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityType;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.component.access.v1.BlockComponentContext;
+import net.fabricmc.fabric.api.component.access.v1.ComponentAccess;
+import net.fabricmc.fabric.api.component.access.v1.ComponentType;
+import net.fabricmc.fabric.api.component.access.v1.EntityComponentContext;
+import net.fabricmc.fabric.api.component.access.v1.ItemComponentContext;
+
+public final class ComponentTypeImpl<T> implements ComponentType<T> {
```

Comment:
I don't think having unchecked casts everywhere is sane. This class should have a field for a `Class<T>` object - would also be useful for introspection.

Reply:
I'm abusing type erasure here to avoid creating separate context/access for each component type. If there's a better way to accomplish that I'm interested.

Reply:
Well I guess there is no helping the type erasure abuse. I do still think a getter for the class object would be helpful for introspection (when you somehow get a `ComponentType<?>`)

Reply:
ComponentType carries the type now so it would be easy to expose it. Will do so.